config/src/main/java/com/typesafe/config/impl/BeanFactory.java

```diff
@@ -0,0 +1,160 @@
+ */
+public class BeanFactory {
+
+    public static <T> T create(Config config, Class<T> clazz) {
+        return createInternal(config, clazz);
+    }
+
+    /**
+     * https://github.com/typesafehub/config/blob/master/HOCON.md#duration-format
+     */
+    private static final Map<TimeUnit,List<String>> TIME_UNITS = new HashMap<TimeUnit,List<String>>();
+    static {
+        TIME_UNITS.put(TimeUnit.NANOSECONDS, Arrays.asList("ns", "nano", "nanos", "nanosecond", "nanoseconds"));
+        TIME_UNITS.put(TimeUnit.MICROSECONDS, Arrays.asList("us", "micro", "micros", "microsecond", "microseconds"));
+        TIME_UNITS.put(TimeUnit.MILLISECONDS,Arrays.asList("ms", "milli", "millis", "millisecond", "milliseconds"));
+        TIME_UNITS.put(TimeUnit.SECONDS,Arrays.asList("s", "second", "seconds"));
+        TIME_UNITS.put(TimeUnit.MINUTES,Arrays.asList("m", "minute", "minutes"));
+        TIME_UNITS.put(TimeUnit.HOURS,Arrays.asList("h", "hour", "hours"));
+        TIME_UNITS.put(TimeUnit.DAYS,Arrays.asList("d", "day", "days"));
+    }
+
+
+    private static <T> T createInternal(Config config, Class<T> clazz) {
+        Map<String, ?> configAsMap = config.root().unwrapped();
+        Map<String, Object> configProps = new HashMap<String, Object>();
+        Map<String,String> originalNames = new HashMap<String, String>();
+        for (Map.Entry<String, ?> configProp : configAsMap.entrySet()) {
+            configProps.put(toCamelCase(configProp.getKey()), configProp.getValue());
+            originalNames.put(toCamelCase(configProp.getKey()),configProp.getKey());
+        }
+
+        try {
+            T bean = clazz.newInstance();
+            for (PropertyDescriptor beanProp : Introspector.getBeanInfo(clazz).getPropertyDescriptors()) {
+                if (beanProp.getReadMethod() == null || beanProp.getWriteMethod() == null) {
+                    continue;
+                }
+                Method setter = beanProp.getWriteMethod();
+                Object configValue = configProps.get(beanProp.getName());
+                if (configValue == null) {
+                    throw new ConfigException.Generic(
+                            "Could not find " + beanProp.getName() + " from " + clazz.getName() + " in config.");
+                }
+                if (configValue instanceof Map) {
+                    configValue = createInternal(config.getConfig(originalNames.get(beanProp.getDisplayName())), beanProp.getPropertyType());
+                } else {
+                    Class parameterClass = setter.getParameterTypes()[0];
+                    configValue = getValueWithAutoConversion(parameterClass, config, originalNames.get(beanProp.getDisplayName()));
+                }
+                setter.invoke(bean, configValue);
+
+            }
+            return bean;
+        } catch (IntrospectionException e) {
+            throw new ConfigException.Generic("Could not resolve a string method name.", e);
```

Comment:
This exception is raised in case when there is a property in ConfigClass that is not present in config file. I.e. we assume that config should set all class properties.
May be it would be better to add some ParsingOptions (for example STRICT(for above behaviour whan we require all props to be present and NON_STRICT when we simply skip non set properties))?